### PR TITLE
Clean Gamerules (2)

### PR DESCRIPTION
**Pull request by bossley9**
## Problem
VRBase has grown into a large behemoth of code with highly extensible members; however, for the scopes of most projects, many of these features are largely unneeded and unused. We need a simpler VRBase set of classes which are easy for new VRBase contributers to understand and provide core functionality that is universal across the majority of VR projects.

In order to make VRBase a plugin, we need to remove all global variables or macros, since they will not translate over to individual project classes due to scope.

In particular, we will not be able to use `Globals` in the future and we need to phase out the code systematically. I believe the best way to go about this is by moving everything to `GameRules` (now renamed to `GameBase` as static members. In addition to controlling game logic flow, `GameBase` will also contain all generic project utilities and "global" variables that may be needed throughout a project.

This PR is one of many that will focus on cleaning existing `GameRules` code and moving global items into `GameRules`.

## Solution
I decided to use a static `TArray` variable to record and count the number of `BaseEntities`. The list is incremented and decremented on each `BaseEntity`'s `BeginPlay` and `EndPlay` methods respectively. I actually think this is a slightly more elegant method of entity counting than the original implementation because

1. [We don't have to worry about duplication](https://docs.unrealengine.com/en-US/Programming/UnrealArchitecture/Actors/ActorLifecycle/index.html), i.e. we don't need to worry about duplicated actors OR things not getting called properly in cooked games.
2. We don't need a separate class (i.e. `CGlobalVars`) to manage resetting the entity counts when the game begins and ends.
3. `BeginPlay` and `EndPlay` are both discouraged from being overridden in all derived IBaseEntity classes, so why not use the empty function implementations?
4. With this implementation, VRBase grows more in similarity to VRBase-Unity. This is a benefit because it makes it much easier for devs to switch between Unreal and Unity.
5. `BeginPlay` and `EndPlay` methods are so much easier to understand than `DestroyEntity`, `RemoveSelfFromLists`, `AddSelfToLists`, `PostDuplicate`, ... if you disagree with me then you're blatantly wrong.

I still have yet to remove the original entity counting system since I want to make sure I do not break anything in our existing projects (i.e. CrimeSolver).

### tl;dr - I reinvented a better entity counting system but I haven't deleted any of the old counting system code yet bc I don't want to break CrimeSolver until I know it's safe to remove it

### CODE REVIEW

File: ABaseEntity/ABaseEntity.cpp [hunk 1]
```diff
@@ -12,7 +12,8 @@ ABaseEntity::ABaseEntity() : IBaseEntity() {
 
 void ABaseEntity::BeginPlay() {
     g_pGlobals->checkReset();
-    AActor::BeginPlay();
```
Comment (bossley9) [inline ABaseEntity/ABaseEntity.cpp:15]: It's not good practice to explicitly state the name of the class. It's a lot better to use `Super`

File: ABasePawn/ABasePawn.h [hunk 2]
```diff
@@ -18,21 +18,29 @@ class VRBASE_API ABasePawn : public APawn, public IBaseEntity {
 public:
     ABasePawn();
 
-    virtual void BeginPlay() override {
```
Comment (bossley9) [inline ABasePawn/ABasePawn.h:21]: Removing `virtual` since overriding should be heavily discouraged

File: ABasePawn/ABasePawn.h [hunk 3]
```diff
@@ -18,21 +18,29 @@ class VRBASE_API ABasePawn : public APawn, public IBaseEntity {
...
             PostInit();
         }
     }
+    void EndPlay(const EEndPlayReason::Type EndPlayReason) override {
+        Super::EndPlay(EndPlayReason);
+        IBaseEntity::EndPlay(EndPlayReason);
+    }
+
     virtual void Tick(float deltaTime) override {}
+
```
Comment (bossley9) [inline ABasePawn/ABasePawn.h:38]: w h i t e s p a c e

File: ABasePawn/ABasePawn.h [hunk 4]
```diff
@@ -18,21 +18,29 @@ class VRBASE_API ABasePawn : public APawn, public IBaseEntity {
...
     virtual void PostDuplicate(EDuplicateMode::Type mode) override {
         Super::PostDuplicate(mode);
         IBaseEntity::PostDuplicate(mode);
     }
+
```
Comment (bossley9) [inline ABasePawn/ABasePawn.h:43]: m o r e   w h i t e s p a c e

File: AGameBase/AGameBase.cpp [hunk 5]
```diff
@@ -45,21 +45,20 @@ void AGameBase::Tick(float deltaTime) {
...
 
-UWorld* g_pWorld = NULL;
-void    AGameBase::BeginPlay() {
+void AGameBase::BeginPlay() {
     Super::BeginPlay();
-    ADebug::Assert((bool)(s_iEntityCount = g_entList.Num()),
-                   "\nIBaseEntity::s_iEntityCount = g_entList.Num()");
+
+    ADebug::Assert(s_iEntityCount == g_entList.Num(),
+                   "\nIBaseEntity::s_iEntityCount == g_entList.Num()");
```
Comment (bossley9) [inline AGameBase/AGameBase.cpp:52]: This assert statement was written incorrectly and I didn't catch it the first time

File: System/Globals.h [hunk 6]
```diff
@@ -25,7 +25,6 @@ class CGlobalVars {
 };
 
 extern CGlobalVars* g_pGlobals;
-extern UWorld*      g_pWorld;
```
Comment (bossley9) [inline System/Globals.h:28]: An old pointer that no longer exists
